feat: payment gateway composition layer

lfmendoza opened the pull request:
## ✨ Descripción

Este Pull Request resuelve problemas de seguridad de tipos en los providers de pagos (PayPal, Visanet, Pagadito) y el servicio de base de datos PostgreSQL, eliminando el uso de tipos `any` y mejorando la robustez del código.

* **Problema Resuelto:**
    * Soluciona errores de linter "Unsafe assignment of an `any` value" en los providers de pagos
    * Elimina el uso de tipos `any` en el manejo de respuestas de APIs externas
    * Mejora la seguridad de tipos en el servicio de base de datos PostgreSQL

* **Impacto:**
    * Mejora la seguridad de tipos en toda la capa de infraestructura de pagos
    * Reduce el riesgo de errores en tiempo de ejecución al acceder a propiedades de objetos
    * Mantiene la funcionalidad existente sin cambios en la API pública

## 🚀 Cambios

* **PayPal Provider (`paypal.provider.ts`):**
    * Implementado manejo seguro de tipos para `error.response?.data` usando type guards
    * Añadido casting explícito de `response.data` a tipos específicos
    * Mejorado el manejo de errores con validación de tipos antes de acceder a propiedades

* **Visanet Provider (`visanet.provider.ts`):**
    * Aplicado el mismo patrón de manejo seguro de tipos que PayPal
    * Eliminado el uso de `error: any` en favor de `error: unknown`
    * Implementado type guards para validar estructura de respuesta antes de acceder a propiedades

* **Pagadito Provider (`pagadito.provider.ts`):**
    * Corregido el manejo de tipos en `getAccessToken()` y `initiatePayment()`
    * Añadido casting seguro para `response.data` en ambos métodos
    * Implementado manejo consistente de errores con type guards

* **PostgreSQL Service (`postgres.service.ts`):**
    * Eliminado el uso de tipos `any` en la propiedad `paymentProviderCredential`
    * Simplificado el servicio para evitar dependencias de tipos no generados
    * Mantenida la funcionalidad base sin comprometer la seguridad de tipos

## 🧪 Pruebas

* **Pruebas de Integración:**
    * Verificar que los providers de pagos manejan correctamente las respuestas de error
    * Probar escenarios donde las APIs externas devuelven diferentes estructuras de respuesta
    * Validar que el servicio de base de datos se inicializa correctamente

* **Pruebas Automatizadas:**
    * [x] Pruebas unitarias actualizadas para validar el manejo seguro de tipos
    * [x] Pruebas de integración para verificar la funcionalidad de los providers
    * [x] Verificación de que no se generan nuevos errores de linter

* **Detalles sobre pruebas específicas ejecutadas:**
    * `npm run lint` - Verificación de que no hay errores de tipos `any`
    * `npm run test` - Ejecución de pruebas unitarias existentes
    * Verificación manual de manejo de errores en providers de pagos

## ✅ Checklist

* [x] He leído la guía [CONTRIBUTING.md](https://github.com/GerardoFdez7/easystore-services/blob/main/CONTRIBUTING.md).
* [x] He realizado una auto-revisión de mi propio código.
* [x] He comentado mi código, particularmente en áreas difíciles de entender.
* [x] Mis cambios no generan nuevas advertencias.
* [x] Todas las pruebas unitarias/de integración existentes y nuevas pasan localmente con mis cambios.
* [x] Se ha respetado la **Arquitectura DDD** (Domain, Application, Infrastructure, Presentation) en la implementación.
* [x] Los patrones **CQRS, Factory y Repository** han sido aplicados correctamente.
* [x] Las transacciones y el manejo de errores han sido considerados y probados.
* [x] Se han considerado las implicaciones de seguridad y rendimiento.

lfmendoza committed:
feat: payment gateway composition layer

Refs: pyaments

GerardoFdez7 commented:
Sigue el patrón de diseño establecido en el repositorio a lo largo de los dominios.

lfmendoza committed:
feat: update payment gateway transactions workflow

lfmendoza committed:
fix: prisma schema generation for payment gateway settings

lfmendoza committed:
fix: eslint errors in prisma type schemas

lfmendoza committed:
refactor: clean up payment gateway module

lfmendoza committed:
feat: improvements payment gateway module

lfmendoza committed:
fix: dependency injection payment providers

lfmendoza committed:
fix: payment providers settings

lfmendoza committed:
refactor: payment validation for transactions, refunds and completed operations

lfmendoza committed:
refactor: encryption for payment credentials

lfmendoza committed:
fix: payment capture end to end validations

lfmendoza committed:
feat: centralized logging audits and payment operations

lfmendoza committed:
fix: replacing winston logger for pino in payment logs

lfmendoza committed:
fix: eslint errors type assertion

lfmendoza committed:
chore: removing payment gateway status doc

lfmendoza committed:
chore: updating project documentation and guidelines

lfmendoza committed:
fix: :bug: revert .env.example file

lfmendoza committed:
fix: adding missing env variables for payment crendtial encryption

GerardoFdez7 commented on src/domains/authentication/infrastructure/jwt/jwt-handler.ts:
Por qué encerrar la variable de jwtSecret en una función?

GerardoFdez7 commented on src/types/cybersource-rest-client.d.ts:
Ya existe una carpeta types en common/types favor mover. Y realmente es necesario este archivo? No se usa en ningún lugar

GerardoFdez7 commented on src/main.ts:
Me parecen innecesarios estos cambios

GerardoFdez7 commented on src/infrastructure/database/postgres.service.ts:
1. Ya no es necesario importar el logger e instanciarlo, está globalmente como los console.logs
2. Me parece más descriptivo usar Postgres literalmente en lugar de Database si en el futuro implementamos otra DB en algún dominio
3. Si estos logs van a estar acá, me parecen innecesario la revisión en main.ts

GerardoFdez7 commented on src/domains/tenant/presentation/graphql/payment-credentials.resolver.ts:
```diff
@@ -0,0 +1,128 @@
+import { GetPaymentCredentialsDto } from '../../application/queries/payment-credentials/get-payment-credentials.dto';
+import { Public } from '@common/decorators';
+
+@Public()
```
Como va a ser Publico el resolver si necesita el id del tenant para todas las peticiones? Es decir, claramente debio haber iniciado sesión primero

GerardoFdez7 commented on src/domains/tenant/presentation/graphql/payment-credentials.resolver.ts:
Utilizar decorador @CurrentUser e importar JwtPayload como en todos los resolvers en el repositorio. Ya que el cliente no almacena ningún id, el backend es el que setea la cookie con todos sus paramentros, y accede a ella en las peticiones que necesite.

GerardoFdez7 commented on src/domains/tenant/infrastructure/persistence/postgres/tenant.repository.ts:
```diff
@@ -25,9 +25,11 @@ export default class TenantRepository implements ITenantRepository {
-            domain: tenantDto.domain,
+            domain:
+              tenantDto.domain ||
+              `${tenantDto.ownerName.toLowerCase()}.easystore.com`,
```
Esto no garantiza que el domain será unico y puede fallar en la creación

GerardoFdez7 commented on src/domains/tenant/infrastructure/persistence/postgres/payment-credentials.repository.ts:
```diff
@@ -0,0 +1,130 @@
+    return count > 0;
+  }
+
+  private mapRecordToEntity(
```
Crear Mapper en application

GerardoFdez7 commented on src/domains/tenant/infrastructure/persistence/postgres/payment-credentials.repository.ts:
No se hace uso de:
- import { PrismaErrorUtils } from '@utils/prisma-error-utils';
- import {
  UniqueConstraintViolationError,
  DatabaseOperationError,
  ResourceNotFoundError,
  ForeignKeyConstraintViolationError,
} from '@shared/errors';

GerardoFdez7 commented on src/domains/payment-gateway/aggregates/entities/index.ts:
Posible dependencia circular al importar everything "*"

GerardoFdez7 commented on src/domains/payment-gateway/aggregates/value-objects/payment/payment-id.vo.ts:
Por qué envolver el value object existente en otro solo de payment que no agrega ninguna funcionalidad especifica de payment? Y además agregar un regex para validarlo cuando id.vo ya cuenta con un metodo create para validarlo correctamente.

GerardoFdez7 commented on src/domains/payment-gateway/aggregates/value-objects/index.ts:
Posible dependencia circular al importar everything "*"

GerardoFdez7 commented on src/domains/payment-gateway/aggregates/entities/payment/payment.attributes.ts:
```diff
@@ -0,0 +1,24 @@
+import { PaymentIdVO } from '../../value-objects/payment/payment-id.vo';
```
Haces un index y los importas por separado

GerardoFdez7 commented on src/domains/payment-gateway/aggregates/entities/payment/payment.entity.ts:
```diff
@@ -0,0 +1,281 @@
+    return new PaymentEntity(attributes);
+  }
+
+  // Getters
```
Estos getters son innecesarios porque puedes acceder a los atributos de forma global, mirar algún dominio para referencia

GerardoFdez7 commented on src/domains/payment-gateway/application/mappers/index.ts:
Posible dependencia circular al importar everything "*"

GerardoFdez7 commented on src/domains/payment-gateway/application/queries/index.ts:
Posible dependencia circular al importar everything "*"

GerardoFdez7 commented on src/domains/payment-gateway/infrastructure/providers/pagadito/pagadito.provider.ts:
Por qué no usar axios aquí tambien?

GerardoFdez7 commented on src/domains/payment-gateway/presentation/graphql/types/index.ts:
Posible dependencia circular al importar everything "*"

GerardoFdez7 commented on src/domains/payment-gateway/presentation/graphql/payment-gateway.resolver.ts:
Utilizar decorador @CurrentUser e importar JwtPayload como en todos los resolvers en el repositorio. Ya que el cliente no almacena ningún id, el backend es el que setea la cookie con todos sus paramentros, y accede a ella en las peticiones que necesite.

GerardoFdez7 commented on src/domains/tenant/application/mappers/tenant/tenant.mapper.ts:
Este cambio es innecesario y rompe con el patrón establecido en el proyecto

GerardoFdez7 reviewed (changes requested):
El dominio rompe con el patrón seguido a lo largo de todos los dominios, creando inconcistencias, posibles errores en tiempo de ejecución, se viola la separación de responsabilidades, y no se sigue el principio KISS. 

Problemas más críticos:
1. No se utiliza zod para realizar las validaciones en value objects, por lo tanto se utiliza class-validator en cada uno de los dtos, lo que rompe el principio DRY y viola DDD completamente al no validar en el dominio las reglas de negocio.
2. Los resolvers no utilizan los decoradores para acceder a los ids de los usuarios que envian las peticiones, de tal forma que el cliente no podrá realizar peticiones.
3. Intenté levantar el dominio con Docker y a pesar de que compila y se supone que expone el puerto correctamente, no me fue posible conectarme como de costumbre por alguna razón:
<img width="894" height="518" alt="image" src="https://github.com/user-attachments/assets/7fb20233-6009-435a-9e02-7d033a3af190" />

GerardoFdez7 committed:
docs: update documentation structure and status

- Remove outdated module references from main README
- Update authentication module status to production ready
- Clean up and consolidate documentation content
